### DISCUSSION TITLE
C1101 with ESP32 initialization fails with return code -2?

### DISCUSSION THREAD

**Discussion by diff-arch** [jgromes/RadioLib#1209] [Q&A] [answered]
Hi,

I'm trying to get a **CC1101** ([868/900MHz](https://quadmeup.com/assets/2017/12/CC1101-868mhz-radio-module-pinout.jpg)) working with an **ESP32 WROOM-32** ([pinout](https://microclub.ch/wp-content/uploads/hm_bbpui/4305/idzowjftqppp447mydc5299y3z64xxjl.png)) over SPI. I'm fairly new to all of this.

Here's how I've currently wired things:
| EPS32 | CC1101 |
| :---      | :---       |
| 3V3     | VCC      |
| GND    | GND     |
| GPIO 23    | MOSI    |
| GPIO 18    | SCLK    |
| GPIO 19    | MISO   |
| GPIO 4    | GBO2   |
| GPIO 2    | GBO0   |
| GPIO 5    | CSN   |

And here's my RadioLib related code:
```c++
#include <Arduino.h>
#include <RadioLib.h>

SPIClass newSPI(VSPI);
CC1101 radio = new Module(5, 2, RADIOLIB_NC, 4, newSPI, SPISettings(2000000, MSBFIRST, SPI_MODE0));

void setup()
{
    Serial.begin(9600);
    Serial.println("Start Setup");

    newSPI.begin();

    const int state = radio.begin();
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println("C1101 initialized successfully");
    }
    else
    {
        Serial.print("C1101 initialization failed, code ");
        Serial.println(state);
        while (true)
        {
            delay(10);
        }
    }
}

void loop()
{
}
```

When I flash the ESP32 with the same code, I get a **return code** of **-2**. I've checked my wiring with a multimeter and everything seems fine.

What am I overlooking here?

Thanks.

### EDIT 1:
I've tested the CC1101 board with my Flipper Zero and it gets instantly recognized as *external* device and I can send data through it. The only difference concerning the wiring is that the GBO2 doesn't get wired to the GPIO.
With the ESP32 I still get a state of -2.

**Comment by jgromes** [comment, marked as answer]
You can start by enabling the [debug mode](https://github.com/jgromes/RadioLib/wiki/Debug-mode), you will most likely see that the device isn't returning any data, pointing to an issue with the SPI. You should double-check the SPI pins; you haven't provided any pin numbers for `newSPI.begin();`, so I'm guessing it will use some defaults - are you sure they are the ones you expect?

**Reply by diff-arch** [reply]
I've enabled "basic debugging" and here's the output:

```
RadioLib Info
Version:  "6.6.0.0"
Platform: "ESP32"
Compiled: "Sep 12 2024" "09:37:25"
RLB_DBG: CC1101 not found! (1 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (2 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (3 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (4 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (5 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (6 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (7 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (8 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (9 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: CC1101 not found! (10 of 10 tries) RADIOLIB_CC1101_REG_VERSION == 0x0000, expected 0x0004/0x0014
RLB_DBG: No CC1101 found!
failed: code -2
``` 

It is as you suspected. Please note that I've since also specified the SPI pins as follows without success:

```c++
#include <Arduino.h>
#include <RadioLib.h>
#include <SPI.h>

constexpr uint8_t SCLK_PIN = 18;  // active-low chip select signal
constexpr uint8_t MISO_PIN = 19;  // main in, sub out
constexpr uint8_t MOSI_PIN = 23;  // main out, sub in
constexpr uint8_t CS_PIN = 5;    // chip select

constexpr uint8_t GDO0_PIN = 2;
constexpr uint8_t GDO2_PIN = 4;

SPIClass spi(VSPI);  // SPI bus attached to pins 5, 18, 19, 23
SPISettings spiSettings(2000000, MSBFIRST, SPI_MODE0);

CC1101 radio = new Module(CS_PIN, GDO0_PIN, RADIOLIB_NC, GDO2_PIN, spi, spiSettings);

void setup()
{
    Serial.begin(115200);

    Serial.print(F("[CC1101] Initializing ... "));

    spi.begin(SCLK_PIN, MISO_PIN, MOSI_PIN, CS_PIN);

    const int state = radio.begin();
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println(F("success!"));
    }
    else
    {
        Serial.print(F("failed: code "));
        Serial.println(state);
        while (true)
        {
            delay(10);
        }
    }
}

void loop()
{
}
```

I've tried verifying the SPI pins by simply printing them. 

```c++
Serial.begin(115200);
Serial.print("MOSI: ");
Serial.println(MOSI);
Serial.print("MISO: ");
Serial.println(MISO);
Serial.print("SCK: ");
Serial.println(SCK);
Serial.print("SS: ");
Serial.println(SS);  
```

I don't know whether that's a valid method or not, but here's the output:

```
MOSI: 23
MISO: 19
SCK: 18
SS: 5
```

Thank you very much for your reply.

**Reply by jgromes** [reply]
It seems like the SPI is not receiving anything at all, are you sure the CC1101 is powered? Other than that, a wiring issue could explain this.

**Reply by diff-arch** [reply]
It was a wiring issue. What had happened is that at first I did check all the cables and connections with a multi-meter, but right after a cable must have slipped out of its metal JST connector sleeve. That was the issue. Now I get a connection.
Thanks for your help so far.

Now, I have a rather long string of raw data (e.g. `... -10164 1101 -404 53 -320 105 -372 53 -352 103 -162 81 -142 169 -666 101 -372 335 -240 ...`) that I want to broadcast over 868.35MHz.
I already know how to set the frequency, but I'm unsure how to send the data. Should I simply parse the string to an `int` array and `digitalWrite` each individual `int` to `GDO0_PIN` with a certain delay between writes?

I've captured this signal with my Flipper Zero.

**Reply by jgromes** [reply]
> string of raw data

What sort of data? Decoded from some modulated signal? A record of signal strength? Raw A/D samples? I think you will need to do some reading on RF signals in general, how are they modulated and encoded before you try to replay a recorded signal.

> digitalWrite each individual int

`digitalWrite` sets an output pin to either logical high, or logical low - that's why it's called *digital*Write. It cannot "output" an integer number.

**Reply by diff-arch** [reply]
> What sort of data?

The recorded raw data looks exactly like the example I've given above, just a lot longer. It's a recording of a remote.
The modulation needs to be 2-FSK, the data rate (kBaud) 512, and data format in asynchronous serial mode.

> digitalWrite sets an output pin to either logical high, or logical low - that's why it's called digitalWrite. It cannot "output" an integer number.

Noted. Do the signed integers maybe represent a delay and their sign a HIGH or LOW state of the pin?

Example:
```
int raw_data[] = {73, -6248, 81, -2096, 103, -300, 1160};
for (int i = 0; i < 7; ++i)
{
  int state = 1;
  int delay = raw_data[i];
  
  if (delay < 0)
    state = 0;  // no sending

  digitalWrite(GDO0_PIN, state);
}
```

**Reply by jgromes** [reply]
> Do the signed integers maybe represent a delay and their sign a HIGH or LOW state of the pin?

I don't know what those numbers represent, but that seems unlikely.

**Reply by diff-arch** [reply]
OK, thanks anyway. I'll investigate it and share what I find out.
Thanks for your patience and kind help.

**Reply by diff-arch** [reply]
> I don't know what those numbers represent, but that seems unlikely.

It turns out it's similar to what I imagined before. The sign indeed represents the pin state (- = LOW, + = HIGH) of transmitting pin, but the unsigned integer value represents the timing between flips (i.e. from low to high or vis versa).

Example:
raw data: 2 -3 6
```
__      ______
   |___|
 2  -3     6  
 ```

**Reply by diff-arch** [reply]
Hi @jgromes,

I've made it this far and the CC1101 is indeed broadcasting over 868.35 MHz, but the remote button behavior is not "emulated" yet.

```c++
#include <Arduino.h>
#include <RadioLib.h>
#include <SPI.h>

#include <vector>
#include <sstream>

#include "raw_data.h"

constexpr uint8_t SCLK_PIN = 18;  // active-low chip select signal
constexpr uint8_t MISO_PIN = 19;  // main in, sub out
constexpr uint8_t MOSI_PIN = 23;  // main out, sub in
constexpr uint8_t CS_PIN = 5;    // chip select

constexpr uint8_t GDO0_PIN = 2;
constexpr uint8_t GDO2_PIN = 4;
constexpr float CC1101_FREQ = 868.35;

SPIClass spi(VSPI);  // SPI bus attached to pins 5, 18, 19, 23
SPISettings spiSettings(2000000, MSBFIRST, SPI_MODE0);

CC1101 radio = new Module(CS_PIN, GDO0_PIN, RADIOLIB_NC, GDO2_PIN, spi, spiSettings);

void setup()
{
    Serial.begin(115200);
    Serial.print(F("[CC1101] Initializing ... "));

    spi.begin(SCLK_PIN, MISO_PIN, MOSI_PIN, CS_PIN);

    int state = radio.begin(CC1101_FREQ); 
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println(F("success!"));
    }
    else
    {
        Serial.print(F("failed: code "));
        Serial.println(state);
        while (true);
    }

    Serial.println(F("Parsing raw data ..."));
    std::istringstream iss(RAW_DATA_STR);
    std::vector<int> numbers;
    int number;

    while (iss >> number)
    {
        numbers.push_back(number);
    }

    // Set the modulation
    Serial.print(F("[CC1101] Setting modulation to OOK ... "));
    state = radio.setOOK(true);
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println(F("success!"));
    }
    else
    {
        Serial.print("failed: code ");
        Serial.println(state);
        while (true);
    }

    // Set the data rate
    Serial.print(F("[CC1101] Setting the bit rate to 512 kbps ... "));
    state = radio.setBitRate(512);
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println(F("success!"));
    }
    else
    {
        Serial.print("failed: code ");
        Serial.println(state);
        while (true);
    }

    // Enable direct asynchronous mode for transmission
    Serial.print(F("[CC1101] Enabling direct asynchronous mode ... "));
    state = radio.transmitDirect();
    if (state == RADIOLIB_ERR_NONE)
    {
        Serial.println(F("success!"));
    }
    else
    {
        Serial.print("failed: code ");
        Serial.println(state);
        while (true);
    }

    // Transmit custom asynchronous data via GPIO
    pinMode(GDO0_PIN, OUTPUT);

    Serial.println(F("Transmitting RAW DATA ..."));
    uint8_t count = 1;
    while (count > 0)
    {
        for (const int num : numbers)
        {
            const unsigned int delay = std::abs(num);
            if (num <= 0)
                digitalWrite(GDO0_PIN, LOW);
            else
                digitalWrite(GDO0_PIN, HIGH);
            delayMicroseconds(delay);
        }
        count--;
    }

}

void loop()
{
}
```

Am I maybe overlooking something?

**Reply by diff-arch** [reply]
OK, I got it to work. The problem was that I explicitly set the modulation to OOK, instead of FSK, which is set by default.
Thanks again. RadioLib is a super neat and handy library and indeed rules them all! ;)

**Reply by jgromes** [reply]
Glad you were able to get it to work, and that you find this project useful! :)
